src/toxcore/tcp/server/client.rs

```diff
@@ -60,7 +61,7 @@ impl Client {
             ip_addr,
             port,
             tx,
-            links: [None; 240],
+            links: Mutex::new(Links::new()),
```

Comment:
Can it be just Links, without Mutex?

Reply:
No, because

```
    fn handle_route_request(&self, pk: &PublicKey, packet: &RouteRequest) -> IoFuture<()> {
        let state = self.state.read();
...
        client_a.links().upgrade(b_id_in_client_a);
        client_b.links().upgrade(a_id_in_client_b);
```

We need interior mutability but on the other side https://doc.rust-lang.org/std/cell/struct.RefCell.html#method.borrow_mut  :

> 

Panics if the value is currently borrowed

Reply:
But we can make a single threaded tcp server with no lock at all =)

Reply:
But it seems it can be done separately, can't it? First take client_a and update it, then release client_a and take client_b.

Reply:
How would you guarantee the order?